api/workerapi/workerhttpserver/server.go

```diff
@@ -0,0 +1,169 @@
+package workerhttpserver
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net"
+	"net/http"
+	"time"
+
+	"github.com/gin-contrib/cors"
+	"github.com/gin-gonic/gin"
+	"github.com/iver-wharf/wharf-cmd/pkg/worker"
+	"github.com/iver-wharf/wharf-core/pkg/cacertutil"
+	"github.com/iver-wharf/wharf-core/pkg/ginutil"
+)
+
+type module interface {
+	register(g *gin.RouterGroup)
+}
+
+// Server lets us start an HTTP server asynchronously.
+type Server struct {
+	address      string
+	port         string
+	srv          *http.Server
+	workerServer *workerServer
+	isRunning    bool
+
+	onServeErrorHandler func(error)
+}
+
+// NewServer creates a new server that can be started by calling Start.
+func NewServer(address string, port string, builder worker.Builder) *Server {
+	return &Server{
+		address:      address,
+		port:         port,
+		workerServer: newWorkerServer(builder),
+	}
+}
+
+// SetOnServeErrorHandler sets the handler to call when an error occurs during
+// serving.
+func (s *Server) SetOnServeErrorHandler(onServeErrorHandler func(error)) {
+	s.onServeErrorHandler = onServeErrorHandler
+}
+
+// Serve starts the HTTP server in a goroutine.
+//
+// Also functions as a force-restart by calling ForceStop if the server is
+// already running, followed by attempting to launch it again.
+//
+// To stop the server you may use GracefulStop or ForceStop.
+func (s *Server) Serve() {
+	s.ForceStop()
+
+	if err := applyHTTPClient(); err != nil {
+		s.onServeErrorHandler(err)
+		return
+	}
```

Comment:
Strongly suggest removing this callback functionality / `throw exception` functionality and instead just rely on returning the error.

Reply:
Changed to return error directly where possible. 👍🏻
If you have a suggestion for how to deal with the errors in the goroutines, besides a callback, that'd be neat as I can't think of any better options at the moment :)

Reply:
Yea you could abuse the closure. Example:

```go
func (s *Server) serve(r *gin.Engine) error {
	s.srv = &http.Server{
		Addr:    fmt.Sprintf("%s:%s", s.address, s.port),
		Handler: r,
	}
	ln, err := net.Listen("tcp", s.srv.Addr)
	if err != nil {
		return err
	}
	var serveError error
	var wg sync.WaitGroup
	wg.Add(1)
	go func() {
		defer wg.Done()
		log.Info().Messagef("Listening and serving HTTP on %s", s.srv.Addr)
		s.isRunning = true
		if err := s.srv.Serve(ln); err != nil {
			if errors.Is(err, http.ErrServerClosed) {
				log.Info().Message("Server closed.")
			} else {
				serveError = err
			}
		}
		s.isRunning = false
	}()
	wg.Wait()
	return serveError
}
```

Reply:
Question becomes then why you don't just run it synchronously to begin with, à la:

```go
func (s *Server) serve(r *gin.Engine) error {
	s.srv = &http.Server{
		Addr:    fmt.Sprintf("%s:%s", s.address, s.port),
		Handler: r,
	}
	ln, err := net.Listen("tcp", s.srv.Addr)
	if err != nil {
		return err
	}
	log.Info().Messagef("Listening and serving HTTP on %s", s.srv.Addr)
	s.isRunning = true
	if err := s.srv.Serve(ln); err != nil {
		if !errors.Is(err, http.ErrServerClosed) {
			return err
		}
		log.Info().Message("Server closed.")
	}
	s.isRunning = false
	return nil
}
```

Reply:
I suppose if the HTTP server is only meant to be started after other stuff has finished then it might as well be synchronous, and be run with `gin.Engine.Run`.

For the gRPC server however, which I assume is supposed to be streaming during building, how would doing it this way let us actually get the error whilst remaining asynchronous?

I might just be misunderstanding something though.

Reply:
Both servers are meant to be run at the same time as the build. Preferrably multiplexed like the gRPC implementation in wharf-api.

If you do `go workerserver.Serve()` then problem solved, it runs the API servers in the background. Then we could for example use a `context.Context` to cancel them when the build is done or when the wharf-cmd-aggregator tells it all data has been successfully transferred.